### PR TITLE
[WIP] Adding `qml.iterative_qpe` 

### PR DESCRIPTION
**Pull request by KetpuntoG**
Adding iterative qpe functionality.
Example:

```python

dev = qml.device("default.qubit", shots = 100)

          @qml.qnode(dev)
          def circuit():

            # Initial state
            qml.PauliX(wires = [0])

            # Iterative QPE
            measurements = qml.iterative_qpe(qml.RZ(2., wires = [0]), estimation_wire = [1], iters = 3)

            return [qml.sample(op = meas) for meas in measurements]

```

### CODE REVIEW

File: pennylane/ops/functions/iterative_qpe.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    Args:
+      base (Operator): the phase estimation unitary, specified as an :class:`~.Operator`
+      estimation_wire (Union[Wire, or int]): the wire to be used for the estimation.
+      iters (int): the number of measurements to be performed.
+
+    Returns:
+      list[MidMeasureMP]: the list of measurements performed.
+
+    .. seealso:: :class:`~.QuantumPhaseEstimation`
```
Comment (Jaybsoni) [inline pennylane/ops/functions/iterative_qpe.py:34]:
```suggestion
    .. seealso:: :class:`~.QuantumPhaseEstimation`, :func:`~.measure`
```

File: pennylane/ops/functions/iterative_qpe.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import numpy as np
+
+
+def iterative_qpe(base, estimation_wire, iters):
+    r"""Function that is in charge of adding to the queue the necessary gates to implement iterative QPE.
+    It returns a list of the mid-circuit measurements performed.
+
+    Args:
+      base (Operator): the phase estimation unitary, specified as an :class:`~.Operator`
+      estimation_wire (Union[Wire, or int]): the wire to be used for the estimation.
```
Comment (Jaybsoni) [inline pennylane/ops/functions/iterative_qpe.py:28]:
```suggestion
      estimation_wire (Union[Wire, int, str]): the wire to be used for the estimation.
```

File: pennylane/ops/functions/iterative_qpe.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+          @qml.qnode(dev)
+          def circuit():
+
+            # Initial state
+            qml.PauliX(wires = [0])
+
+            # Iterative QPE
+            measurements = qml.iterative_qpe(qml.RZ(2., wires = [0]), estimation_wire = [1], iters = 3)
+
+            return [qml.sample(op = meas) for meas in measurements]
```
Comment (Jaybsoni) [inline pennylane/ops/functions/iterative_qpe.py:51]:
```suggestion
              # Initial state
              qml.PauliX(wires = [0])

              # Iterative QPE
              measurements = qml.iterative_qpe(qml.RZ(2., wires = [0]), estimation_wire = [1], iters = 3)

              return [qml.sample(op = meas) for meas in measurements]
```

File: pennylane/ops/functions/iterative_qpe.py (hunk 2 above)
Comment (Jaybsoni) [reply pennylane/ops/functions/iterative_qpe.py:28]: The wire label can be a string as well, also I think in your code examples you pass a list here

File: pennylane/ops/functions/iterative_qpe.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    """
+
+    measurements = []
+
+    for i in range(iters):
+        qml.Hadamard(wires=estimation_wire)
+        qml.ctrl(qml.pow(base, z=2 ** (iters - i - 1)), control=estimation_wire)
+
+        for ind, meas in enumerate(measurements):
+            qml.cond(meas == 1, qml.PhaseShift)(-2 * np.pi / 2 ** (ind + 2), wires=estimation_wire)
```
Comment (Jaybsoni) [inline pennylane/ops/functions/iterative_qpe.py:64]: If meas is 0 then it is treated as False, if meas is 1 it is treated as true, no need to check meas == 1.
```suggestion
            qml.cond(meas, qml.PhaseShift)(-2 * np.pi / 2 ** (ind + 2), wires=estimation_wire)
```

File: pennylane/ops/functions/iterative_qpe.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      base (Operator): the phase estimation unitary, specified as an :class:`~.Operator`
+      estimation_wire (Union[Wire, or int]): the wire to be used for the estimation.
+      iters (int): the number of measurements to be performed.
+
+    Returns:
+      list[MidMeasureMP]: the list of measurements performed.
+
+    .. seealso:: :class:`~.QuantumPhaseEstimation`
+
+    **Example**
```
Comment (Jaybsoni) [inline pennylane/ops/functions/iterative_qpe.py:36]: Maybe an advanced example with multiple shots to show users what they expect to see?

File: tests/ops/functions/test_iterative_qpe.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            qml.CNOT(wires=[1, 3])
+            qml.CNOT(wires=[3, 1])
+
+            return qml.expval(qml.PauliZ(0))
+
+        phi = jax.numpy.array(1.0)
+        assert jax.numpy.isclose(jax.grad(circuit)(phi), jax.grad(manual_circuit)(phi))
+
+    @pytest.mark.parametrize("iters", (1, 2, 3, 4))
+    def test_size_return(self, iters):
```
Comment (Jaybsoni) [inline tests/ops/functions/test_iterative_qpe.py:90]: Nice 👍🏼

File: tests/ops/functions/test_iterative_qpe.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+Unit tests for the iterative_qpe function
+"""
+import pytest
+
+import pennylane as qml
+import numpy as np
+import jax
+
+
+class TestResult:
```
Comment (Jaybsoni) [inline tests/ops/functions/test_iterative_qpe.py:24]: We should probably test that it works with all of the measurement processes (expval, probs, etc.) that are supported by mid-circuit measurements.

File: tests/ops/functions/test_iterative_qpe.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    def test_check_gradients(self):
+        """Test to check that the gradients are correct comparing with the expanded circuit"""
+        dev1 = qml.device("default.qubit")
+
+        @qml.qnode(dev1)
+        def circuit(theta):
+            m = qml.iterative_qpe(qml.RZ(theta, wires=[0]), [1], iters=2)
+            return qml.expval(qml.PauliZ(0))
+
+        dev2 = qml.device("default.qubit")
```
Comment (Jaybsoni) [inline tests/ops/functions/test_iterative_qpe.py:68]: you can re-use the same device, don't need to define a new one.

**Review by Jaybsoni — COMMENTED** [review]
Just some comments, I can provide a more detail review once we have more tests 👍🏼

File: tests/ops/functions/test_iterative_qpe.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        sample_list = sample_list.T
+        output = qml.probs().process_samples(np.array(sample_list), wire_order=[0, 1, 2])
+
+        @qml.qnode(dev)
+        def circuit_qpe():
+            # Initial state
+            qml.PauliX(wires=[0])
+
+            # Iterative QPE
+            qml.QuantumPhaseEstimation(qml.RZ(2.0, wires=[0]), estimation_wires=[1, 2, 3])
```
Comment (soranjh) [inline tests/ops/functions/test_iterative_qpe.py:53]: Should be `qml.RZ(phi, wires=[0])` ?

**Review by soranjh — COMMENTED** [review]
Thanks @KetpuntoG, nice PR! We need more tests .

File: tests/ops/functions/test_iterative_qpe.py (hunk 9 above)
Comment (soranjh) [inline tests/ops/functions/test_iterative_qpe.py:53]:
```suggestion
            qml.QuantumPhaseEstimation(qml.RZ(phi, wires=[0]), estimation_wires=[1, 2, 3])
```

File: pennylane/ops/functions/iterative_qpe.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+"""
+This file contains the function which generates the iterative QPE subroutine.
+"""
+
+import pennylane as qml
+import numpy as np
+
+
+def iterative_qpe(base, estimation_wire, iters):
+    r"""Function that is in charge of adding to the queue the necessary gates to implement iterative QPE.
```
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/ops/functions/iterative_qpe.py:23]: Yes please let's make this a user facing description.

File: pennylane/ops/functions/iterative_qpe.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    It returns a list of the mid-circuit measurements performed.
```
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/ops/functions/iterative_qpe.py:24]: Nice! Any paper to link to?

File: pennylane/ops/functions/iterative_qpe.py (hunk 5 above)
Comment (trbromley) [reply pennylane/ops/functions/iterative_qpe.py:36]: Curious to see what the advanced example looks like - also as validation that IQPE works as we expect.

If a more advanced example is too long, it could go into a usage details section. Otherwise, best left to a demo.

Comment (KetpuntoG) [reply pennylane/ops/functions/iterative_qpe.py:36]: The advanced example is not too pretty right now. We would have to wait until we can have generalized statistics on the measurements. However, if needed now, the post-processing would be like this:

```python
dev = qml.device("default.qubit", shots=1000)

@qml.qnode(dev)
def circuit_iterative():
    # Initial state
    qml.PauliX(wires=[0])

    # Iterative QPE
    measurements = qml.iterative_qpe(qml.RZ(phi, wires=[0]), estimation_wire=[1], iters=3)

    return [qml.sample(op=meas) for meas in measurements]

# This is the part that will change hopefully. <--------
sample_list = np.array(circuit_iterative())
sample_list = sample_list.T
output = qml.probs().process_samples(np.array(sample_list), wire_order=[0, 1, 2])

```

**Comment by Jaybsoni** [comment]
Also I think we should add it to this section of the docs: https://docs.pennylane.ai/en/stable/introduction/operations.html#operator-to-other-functions

**Review by trbromley — CHANGES_REQUESTED** [review]
In discussion - but we'll close this PR.
